File: 1696-jump-game-vi/1696-jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [0] * n
        max_value = 0
        heap = []
        heapify(heap)
        for i in range(min(k+1, n)):
            if heap:
                max_value, _ = heap[0]
                max_value = - max_value
            dp[i] += max_value + nums[i]
            heappush(heap,[-dp[i], i])
        for i in range(min(k+1,n), n):
            while heap and heap[0][1] < i - k:
                heappop(heap)
            max_value, _ = heap[0]
            dp[i] = -max_value + nums[i]
            heappush(heap,[-dp[i], i])
        return dp[-1]
            
        

File: 1696-jump-game-vi/1696-jump-game-vi.java
class Solution {
    public int maxResult(int[] nums, int k) {
        int N = nums.length;
        int[] score = new int[N];
        Deque<Integer> queue = new LinkedList<>();
        queue.offerLast(0);
        score[0] = nums[0];
        for(int i = 0; i < N - 1; i++){
            //update the queue
            while(!queue.isEmpty() && i - queue.peekFirst() >= k) queue.pollFirst();
            while(!queue.isEmpty() && score[queue.peekLast()] <= score[i]) queue.pollLast();
            queue.offerLast(i);
            score[i+1] = score[queue.peekFirst()] + nums[i+1];
        }
        return score[N-1];
        
    }
}